flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/functions/table/fullcache/inputformat/InputFormatCacheLoaderTest.java

```diff
@@ -149,19 +155,33 @@ void testCloseAndInterruptDuringReload() throws Exception {
         Future<?> future = executorService.submit(cacheLoader);
         executorService.shutdownNow(); // internally interrupts a thread
         assertThatNoException().isThrownBy(future::get); // wait for the end
-        // check that we didn't process all elements, but reacted on interruption
-        assertThat(sleepCounter).hasValueLessThan(totalSleepCount);
         assertThat(metricGroup.numLoadFailuresCounter.getCount()).isEqualTo(0);
+    }
 
-        sleepCounter.set(0);
+    @Test
+    void testCloseDuringReload() throws Exception {
+        AtomicInteger recordsCounter = new AtomicInteger(0);
+        int totalRecords = TestCacheLoader.DATA.values().stream().mapToInt(Collection::size).sum();
+        CountDownLatch latch = new CountDownLatch(1);
+        Runnable reloadAction =
+                ThrowingRunnable.unchecked(
+                        () -> {
+                            recordsCounter.incrementAndGet();
+                            latch.await();
```

Comment:
Doing that leads to a deadlock. Even just adding `Thread.sleep` shortly before closing the cache loader makes the test run into a deadlock because [CacheLoader:101](https://github.com/apache/flink/blob/340b100f2de5e0d90ba475aa8a00e359a61442ce/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/CacheLoader.java#L101) acquires the `CacheLoader.reloadLock` and initiates the reloading of the cache synchronously while having the lock acquired. The data reload blocks due to the `CountDownLatch` we use in the test. The subsequent call of close triggers the shutdown of the `InputSplitCacheLoadTask` instances through `InputSplitCacheLoadTask#stopRunning`. But the shutdown of those tasks will never be done because the data loading is still blocked and the tasks are stuck in the [while loop](https://github.com/apache/flink/blob/340b100f2de5e0d90ba475aa8a00e359a61442ce/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/inputformat/InputSplitCacheLoadTask.java#L71). [InputFormatCacheLoader#close](https://github.com/apache/flink/blob/a024a366f73f822bb4fd35db737ac2b8177f6b25/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/inputformat/InputFormatCacheLoader.java#L129) will continue with calling `CacheLoader#close` which tries to acquire the lock that's already acquired by the thread that executes `InputFormatCacheLoader` and we end up in a deadlock.

My suspicion is that the test as it is pushed right now is passing because we're calling close before the data reload is initiated. AFAIU, that's not what we want to test, is it? Please correct me if I'm wrong here.